api/src/services/authentication.ts

```diff
@@ -131,16 +142,12 @@ export class AuthenticationService {
 		}
 
 		if (password !== undefined) {
-			if (!user.password) {
-				emitStatus('fail');
-				await stall(STALL_TIME, timeStart);
-				throw new InvalidCredentialsException();
-			}
-
-			if ((await argon2.verify(user.password, password)) === false) {
+			try {
+				await provider.verify({ ...user }, password);
+			} catch (e) {
```

Comment:
This is the only bit I'm not sure about yet. The password check is pretty proprietary to one or to drivers.

To me, it looks like this _authenticate_ method has 3 "parts", 2 of which are shared between all drivers:

1) Before-login attempt. Tracks a rate limiter point, and checks if the user record exists and is active
2) Proprietary check from the driver. In the case of the local driver, it verifies the password
3) Start the session, and return the tokens

I'm wondering if we should split up this authenticate method into 2 methods: `validateLoginAttempt` and `login`.  These would map to no 1 and 3 respectively. That would mean that it's up to the driver implementation to decide whether any other validation is required, like checking a hash string for password validity 🤔

Reply:
Hm... I really couldn't say. I think we'll learn pretty quickly once we start adding new auth flows how the authenticate function could be improved, but right now it's hard for me to visualize the different scenarios.

I can, however, see the value in adding a `validateLoginAttempt` function to the auth intereface to compliment the `validate` function, since validating a password and validating a login aren't always the same thing.